rak811.go

```diff
@@ -42,98 +43,117 @@ func New(conf *serial.Config) (*Lora, error) {
 	}
 
 	return &Lora{
-		port: port,
+		port:    port,
+		timeout: defaultConfig.ReadTimeout,
 	}, nil
 }
 
-func (l *Lora) tx(cmd string) (string, error) {
+func (l *Lora) txr(cmd string, lines int) (string, error) {
 	if _, err := l.port.Write(createCmd(cmd)); err != nil {
-		return "", fmt.Errorf("failed to write command %q with: %s", cmd, err)
+		return "", errors.Wrapf(err, "writing a command:%v", cmd)
 	}
+	return l.tr(lines)
+}
 
-	scanner := bufio.NewScanner(bufio.NewReader(l.port))
+func (l *Lora) tr(lines int) (string, error) {
+	reader := bufio.NewReader(l.port)
 	var resp string
-	for scanner.Scan() {
-		resp += scanner.Text()
-	}
-
-	if err := scanner.Err(); err != nil {
-		return "", fmt.Errorf("failed reading response: %s", err)
+	start := time.Now()
+	for line := 0; line < lines; line++ {
+		for {
+			r, err := reader.ReadString('\n')
+
+			if err != nil {
+				if err == io.EOF { // The serial port has a max timeout of 25sec so we rely on the l.timeout.
+					continue
+				}
+				return "", errors.Wrap(err, "failed read")
+			}
+			if strings.HasPrefix(r, "ERROR") {
+				return "", errors.Errorf(r)
+			}
+			resp += r
+			if time.Since(start) > l.timeout {
+				return "", errors.Errorf("no response within:%v", l.timeout)
+			}
+			break
+		}
 	}
-	return resp, nil
+	return strings.TrimSuffix(strings.TrimSuffix(strings.TrimSpace(resp), "\r"), "\n"), nil
 }
 
 //
 // System Commands
 //
 
-// Version get module version
+// Version get module version.
 func (l *Lora) Version() (string, error) {
-	return l.tx("version")
+	return l.txr("version", 1)
 }
 
-// Sleep enter sleep mode
+// Sleep enter sleep mode.
 func (l *Lora) Sleep() (string, error) {
-	return l.tx("sleep")
+	return l.txr("sleep", 1)
+}
+
+// Wakeup to wake up the module.
+func (l *Lora) Wakeup() (string, error) {
+	return l.txr("wake", 1) // Any command will wake it up.
 }
 
-// Reset module or LoRaWAN stack
-// 0: reset and restart module
+// Reset module or LoRaWAN stack.
+// 0: reset and restart module,
 // 1: reset LoRaWAN stack and the module will reload
-// LoRa configuration from EEPROM
+// LoRa configuration from EEPROM.
 func (l *Lora) Reset(mode int) (string, error) {
-	return l.tx(fmt.Sprintf("reset=%d", mode))
+	return l.txr(fmt.Sprintf("reset=%d", mode), 4)
 }
 
 // HardReset the module by reseting the hat pins.
 func (l *Lora) HardReset() (string, error) {
 	pin, err := rpi.OpenPin(rpi.GPIO17, gpio.ModeOutput)
 	if err != nil {
-		return "", fmt.Errorf("error opening pin err:%v", err)
+		return "", errors.Wrapf(err, "opening pin")
```

Comment:
let's keep the standard library and prepare for the upcoming changes in 1.13.

Reply:
errors.Wrapf has been around for a while? What is the problem to use it here?

Reply:
Yea I know but is a dependency and is not required. Check the link I sent you the other review about the error wrapping coming in 1.13

Reply:
what is the difference if you depend on fmt or errors?

Reply:
Well errors is an external dependency `"github.com/pkg/errors"` and is not required, besides to be consistent with the all others errors.

Reply:
not that important so reverted to unblock the PR.